# Telco Customer Churn — Feature Engineering

## Goal
Prepare the data for modeling:
- Drop unnecessary columns
- Encode categorical variables
- Handle class imbalance
- Save processed data

## 1. Imports & Load Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder

pd.set_option('display.max_columns', None)

In [3]:
df = pd.read_csv('../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')
print(df.shape)
df.head()

(7043, 21)


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 2. Drop Unnecessary Columns

In [4]:
# customerID is just an identifier — no predictive value
df = df.drop(columns=['customerID'])
print(f"Shape after drop: {df.shape}")

Shape after drop: (7043, 20)


## 3. Encode Target Variable

In [6]:
# Churn: Yes=1, No=0
df['Churn'] = (df['Churn'] == 'Yes').astype(int)
print(df['Churn'].value_counts())

Churn
0    5174
1    1869
Name: count, dtype: int64


## 4. Encode Categorical Variables

In [7]:
# Fix TotalCharges — stored as object
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df['TotalCharges'] = df['TotalCharges'].fillna(0)

In [8]:
# Binary: Yes=1, No=0
binary_cols = ['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling']
for col in binary_cols:
    df[col] = (df[col] == 'Yes').astype(int)

# gender: Male=1, Female=0
df['gender'] = (df['gender'] == 'Male').astype(int)

# Churn: Yes=1, No=0
df['Churn'] = (df['Churn'] == 'Yes').astype(int)

print("Binary encoding done")

Binary encoding done


In [9]:
# These columns have 3 values including 'No internet service' / 'No phone service'
# Treat 'No internet service' and 'No phone service' same as 'No' → 0
three_val_cols = ['MultipleLines', 'OnlineSecurity', 'OnlineBackup',
                  'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']

for col in three_val_cols:
    df[col] = (df[col] == 'Yes').astype(int)

print("Three-value encoding done ")

Three-value encoding done 


In [ ]:
# One-Hot Encoding for multi-class columns 
df = pd.get_dummies(df, columns=['InternetService', 'Contract', 'PaymentMethod'], drop_first=False)

print(f"Shape after encoding: {df.shape}")
df.head()

Shape after encoding: (7043, 27)


,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,PaperlessBilling,MonthlyCharges,TotalCharges,Churn,InternetService_DSL,InternetService_Fiber optic,InternetService_No,Contract_Month-to-month,Contract_One year,Contract_Two year,PaymentMethod_Bank transfer (automatic),PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
0,0,0,1,0,1,0,0,0,1,0,0,0,0,1,29.85,29.85,0,True,False,False,True,False,False,False,False,True,False
1,1,0,0,0,34,1,0,1,0,1,0,0,0,0,56.95,1889.50,0,True,False,False,False,True,False,False,False,False,True
2,1,0,0,0,2,1,0,1,1,0,0,0,0,1,53.85,108.15,0,True,False,False,True,False,False,False,False,False,True
3,1,0,0,0,45,0,0,1,0,1,1,0,0,0,42.30,1840.75,0,True,False,False,False,True,False,True,False,False,False
4,0,0,0,0,2,1,0,0,0,0,0,0,0,1,70.70,151.65,0,False,True,False,True,False,False,False,False,True,False


In [11]:
# Verify no object columns remain
remaining = df.select_dtypes(include='object').columns.tolist()
print(f"Remaining object columns: {remaining}")

Remaining object columns: []


## 5. Save Processed Data

In [12]:
df.to_csv('../data/processed/telco_processed.csv', index=False)
print("Processed data saved ")
print(f"Final shape: {df.shape}")
df.head()

Processed data saved 
Final shape: (7043, 27)


,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,PaperlessBilling,MonthlyCharges,TotalCharges,Churn,InternetService_DSL,InternetService_Fiber optic,InternetService_No,Contract_Month-to-month,Contract_One year,Contract_Two year,PaymentMethod_Bank transfer (automatic),PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
0,0,0,1,0,1,0,0,0,1,0,0,0,0,1,29.85,29.85,0,True,False,False,True,False,False,False,False,True,False
1,1,0,0,0,34,1,0,1,0,1,0,0,0,0,56.95,1889.50,0,True,False,False,False,True,False,False,False,False,True
2,1,0,0,0,2,1,0,1,1,0,0,0,0,1,53.85,108.15,0,True,False,False,True,False,False,False,False,False,True
3,1,0,0,0,45,0,0,1,0,1,1,0,0,0,42.30,1840.75,0,True,False,False,False,True,False,True,False,False,False
4,0,0,0,0,2,1,0,0,0,0,0,0,0,1,70.70,151.65,0,False,True,False,True,False,False,False,False,True,False


## 6. Conclusions

- Dropped `customerID` — no predictive value
- Fixed `TotalCharges` — converted from string to numeric
- Binary encoded: `gender`, `Partner`, `Dependents`, `PhoneService`, `PaperlessBilling`
- Three-value encoded: `MultipleLines`, `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`
- One-Hot encoded: `InternetService`, `Contract`, `PaymentMethod`
- Final dataset: **7,043 rows × 27 columns** — ready for modeling 